Added support for transferring incoming file shares.

ipasanec opened the pull request:
Fix https://github.com/nextcloud/server/issues/21856

@PVince81 

- new option `--transfer-incoming-shares=1 | 0`
- new config.php option ~~`'transfer-incoming-shares' => true | false`~~ `'transferIncomingShares' => true | false`

The command line option overrules the config.php option.

miaulalala commented on apps/files/lib/Command/TransferOwnership.php:
```diff
@@ -111,12 +123,38 @@ protected function execute(InputInterface $input, OutputInterface $output): int
+					} else if (gettype($includeIncoming) === 'NULL') {
+						$includeIncoming = false;
+					}
+					break;				
```
If you assume `false` you can avoid this case

ipasanec replied:
I changed the code. I hope that I understood everything right.
But at least the code is shorter with same behaviour.

miaulalala commented on apps/files/lib/Service/OwnershipTransferService.php:
```diff
@@ -288,6 +312,33 @@ private function collectUsersShares(string $sourceUid,
+	private function collectIncomingShares(string $sourceUid,
+										OutputInterface $output,
+										View $view
+										/* string $path */): array {
```
If you don't need `$path` any more, you should remove it.

ipasanec replied:
I removed it.
Thanks for the review!

ipasanec commented on apps/files/lib/Command/TransferOwnership.php:
```diff
@@ -83,6 +90,11 @@ protected function configure() {
+				'transfer-incoming-shares',
+				null,
+				InputOption::VALUE_REQUIRED,
+				'transfer incoming shares to destination user'
```
I changed it.
I clarified in the description that only file shares are transferred.

miaulalala replied:
I saw that you changed line 95 to `false`. I'm sorry for confusing you here! - this should stay null, but you can add an additional parameter at the end of `addOption` to declare a default value that gets passed if the user does not specify a value, so you don't need to cover a `NULL` case at all.

So your code would look like this:

```php
)->addOption(
    'transfer-incoming-shares',
    null,
    InputOption::VALUE_OPTIONAL,
    'transfer incoming shares to destination user',
    false
);

```

ipasanec replied:
Learned something today :-)
I changed the default value to `'2'`. Because `'0' == false` the third case would never occur with `false`.

miaulalala replied:
Why do you need the third option?

ipasanec replied:
The third option is for looking up the config.php to define standard behaviour when no option is given.
Of course, it is debatable if this functionality should stay there. (originally it was a customer requirement)

miaulalala replied:
Ah makes sense! :+1:

PVince81 commented on apps/files/lib/Service/OwnershipTransferService.php:
```diff
@@ -356,4 +406,64 @@ private function restoreShares(string $sourceUid,
+					$sameFileSharedWithDestination = false;
+					$shareDeleted = false;
+					foreach ($destinationShares as $key => $destinationShare) {
+						if ($share->getNodeId() === $destinationShare->getNodeId()) {
```
I wonder if we should have used the node id as key for the destination shares for faster lookup ?

this would remove the need to iterate over all destination shares several times and to remove entries in the middle

ipasanec replied:
I changed it accordingly.

PVince81 commented on apps/files/lib/Service/OwnershipTransferService.php:
```diff
@@ -356,4 +406,64 @@ private function restoreShares(string $sourceUid,
+					foreach ($destinationShares as $key => $destinationShare) {
+						if ($share->getNodeId() === $destinationShare->getNodeId()) {
+							$sameFileSharedWithDestination = true;
+							if ($destinationShare->getPermissions() < $share->getPermissions()) {
```
maybe add code comment: "keep the share which has the most permissions and discard the other one"

ipasanec replied:
Done.

PVince81 commented on apps/files/lib/Service/OwnershipTransferService.php:
```diff
@@ -356,4 +406,64 @@ private function restoreShares(string $sourceUid,
+
+					// trigger refetching of the node so that the new owner and mountpoint are taken into account
+					// otherwise the checks on the share update will fail due to the original node not being available in the new user scope
+					$this->userMountCache->clear();
```
I guess you could move this into the if condition below as we only need to refetch the node id when reassigning

ipasanec replied:
I did it somehow. Due to the changed code structure it appears now two times in the code.

PVince81 commented on apps/files/lib/Service/OwnershipTransferService.php:
```diff
@@ -356,4 +406,64 @@ private function restoreShares(string $sourceUid,
+						}
+					}
+
+					if ($sameFileSharedWithDestination === false) {
```
if "$shareDeleted" is true, do we still need to come here ?

ipasanec replied:
I think the new code structure resolved this.

PVince81 reviewed (commented):
Had a deeper look at the logic and added some comments

ipasanec commented on apps/files/lib/Service/OwnershipTransferService.php:
```diff
@@ -288,6 +312,32 @@ private function collectUsersShares(string $sourceUid,
+				break;
+			}
+
+			$shares = array_merge($shares, $sharePage);
```
I changed the assembly of the array. Maybe I shoud change it for the original code too?

PVince81 replied:
as you like. it's fine to leave it

ipasanec commented on apps/files/lib/Service/OwnershipTransferService.php:
```diff
@@ -356,4 +406,64 @@ private function restoreShares(string $sourceUid,
+								   string $destinationUid,
+								   array $sourceShares,
+								   array $destinationShares,
+								   OutputInterface $output) {
```
Oops, done.

ipasanec commented on apps/files/lib/Service/OwnershipTransferService.php:
```diff
@@ -288,6 +312,32 @@ private function collectUsersShares(string $sourceUid,
+
+		$offset = 0;
+		while (true) {
+			$sharePage = $this->shareManager->getSharedWith($sourceUid, IShare::TYPE_USER, null, 50, $offset);
```
I nam uncertain how to document this. Do you mean a change of the command name, like `--transfer-incoming-usershares`?

PVince81 replied:
you could simply add a hint in the help text from the command's help

ipasanec replied:
I changed the help text of the command and of the option.

PVince81 commented on apps/files/lib/Service/OwnershipTransferService.php:
```diff
@@ -22,13 +22,14 @@
  * This program is distributed in the hope that it will be useful,
  * but WITHOUT ANY WARRANTY; without even the implied warranty of
- * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
```
oops, sneaky space

PVince81 commented on apps/files/lib/Service/OwnershipTransferService.php:
```diff
@@ -288,6 +313,42 @@ private function collectUsersShares(string $sourceUid,
+	private function collectIncomingShares(string $sourceUid,
+										OutputInterface $output,
+										View $view,
+										$addKeys = false): array {
```
please add type to make the bot happy

PVince81 commented on apps/files/lib/Service/OwnershipTransferService.php:
```diff
@@ -356,4 +417,52 @@ private function restoreShares(string $sourceUid,
+					$this->shareManager->deleteShare($share);
+				} else if (isset($destinationShares[$share->getNodeId()])) {
+					$destinationShare = $destinationShares[$share->getNodeId()];
+					// Keep the share which has the most permissions and discard the other one.
```
I was not able to think of a scenario where a user is receiving the same node through different **user** shares.
It is usually possible by having a user share and group share with different permissions.

Did you have a scenario in mind ? Otherwise I'm not sure if we need this code, but keeping it doesn't hurt.

PVince81 commented:
- [ ] BUG: must not transfer incoming share if not in the requested path:
    - user2 shares "test" with user1
    - user1 creates dir "sub"
    - `occ files:transfer-ownership user1 user2 --path=sub -vvv --transfer-incoming-shares=1`
    - expected: "test" stays
    - actual: "test" was transferred even though it was not in "sub"

- [ ] BUG: transferred incoming share doesn't stay in folder
    - user2 shares "test" with user1
    - user1 creates dir "sub"
    - user1 moves "test" into "sub"
    - `occ files:transfer-ownership user1 user2 --path=sub -vvv --transfer-incoming-shares=1`
    - expected: "test" is inside "sub" for user2
    - actual: "test" is outside of "sub" and even outside the transferred folder

PVince81 reviewed (changes requested):
Thanks for the changes so far.
I had a few more comments after doing some tests.

See test results above.
I also took the liberty to push an integration test to your branch for the reported cases :smile:

ipasanec commented on apps/files/lib/Service/OwnershipTransferService.php:
```diff
@@ -288,6 +313,42 @@ private function collectUsersShares(string $sourceUid,
+				}
+			} else {
+				foreach ($sharePage as $singleShare) {
+					$shares[] = $singleShare;
```
I had the fantas that it might be beneficial to performance(?)

PVince81 replied:
Good question. I did a quick search but didn't find any post that didn't say "I didn't read the PHP engine code, but...".

I guess you can leave it, it doesn't hurt readibility

PVince81 reviewed (approved):
:+1: I hereby approve the code from @ipasanec 

second reviewer should also check my commits

ipasanec committed:
Added support for transfering incoming file shares.

- new option --transfer-incoming-shares=1 | 0
- new config.php option 'transfer-incoming-shares' => true | false

The command line option overrules the config.php option.

Signed-off-by: Vincent Petry <vincent@nextcloud.com>

PVince81 committed:
Test for transfer ownership with incoming shares

Signed-off-by: Vincent Petry <vincent@nextcloud.com>

PVince81 commented:
/backport to stable22

acsfer commented:
Shouldn't be this documented too? 🤔

PVince81 commented:
I'll take care:

- [ ] update admin_docs for occ command
- [ ] add new setting in config.sample.php

PVince81 commented:
docs done!

Vilasamuni commented:
I've been looking for documentation about how to use this flag and it's not clear to me what the values mean. I've seen `--transfer-incoming-shares=1` and `--transfer-incoming-shares=2` but I haven't found any explanation of what `1` and `2` actually do.

I've got backups, but I'm wary about running commands on real-world data with flags that I don't understand.

Short term request: please explain to me what they do.
Medium term suggestion: update the documentation to make this clearer.

ChristophWurst commented:
> Short term request: please explain to me what they do.

https://github.com/nextcloud/server/pull/28118/files#diff-e299680f9e139f82c630e100e9a4750851e51598a493f185ce8fc7baf2544ea3R127-R147



> Medium term suggestion: update the documentation to make this clearer.

Please open a ticket at https://github.com/nextcloud/documentation/issues/new

ipasanec commented:
The basic usage is:
--transfer-incoming-shares=1 for transferring incoming shares explicitly and 0 for not transferring them explicitly.
This command line option overwrites the value of 'transferIncomingShares' in the config.php if set.
If nothing is set, neither in command line nor in config.php, incoming shares are not transferred.

The default value of '2' for an unset option is necessary to detect a missing value in the command line and give an error message to notify the user.
The only valid values for --transfer-incoming-shares are 0 or 1.

btw. there is no --include-incoming-shares option. (It was changed in the original project, hence some variables use 'include' in the name)
